pulsar-connectors/google-cloud/src/main/java/org/apahce/pulsar/common/io/cloud/gcs/GcsPath.java

```diff
@@ -0,0 +1,627 @@
+     * to a Unix path that does not begin with the root marker (a slash).
+     * GCS has different naming constraints and APIs for working with buckets and
+     * objects, so these two concepts are kept separate to avoid accidental
+     * attempts to treat objects as buckets, or vice versa, as much as possible.
+     *
+     * <p>A GcsPath without an object name is a bucket reference.
+     * A bucket is always a directory, which could be used to lookup or add
+     * files to a bucket, but could not be opened as a file.
+     *
+     * <p>A GcsPath containing neither bucket or object names is treated as
+     * the root of the GCS filesystem.  A listing on the root element would return
+     * the buckets available to the user.
+     *
+     * <p>If {@code null} is passed as either parameter, it is converted to an
+     * empty string internally for consistency.  There is no distinction between
+     * an empty string and a {@code null}, as neither are allowed by GCS.
+     *
+     * @param bucket a GCS bucket name, or none ({@code null} or an empty string)
+     *               if the object is not associated with a bucket
+     *               (e.g. relative paths or the root node).
+     * @param object a GCS object path, or none ({@code null} or an empty string)
+     *               for no object.
+     */
+    public static GcsPath fromComponents(@Nullable String bucket,
+                                         @Nullable String object) {
+        return new GcsPath(null, bucket, object);
+    }
+
+    @Nullable
+    private transient FileSystem fs;
+    @Nonnull
+    private final String bucket;
+    @Nonnull
+    private final String object;
+
+    /**
+     * Constructs a GcsPath.
+     *
+     * @param fs the associated FileSystem, if any
+     * @param bucket the associated bucket, or none ({@code null} or an empty
+     *               string) for a relative path component
+     * @param object the object, which is a fully-qualified object name if bucket
+     *               was also provided, or none ({@code null} or an empty string)
+     *               for no object
+     * @throws java.lang.IllegalArgumentException if the bucket of object names
+     *         are invalid.
+     */
+    public GcsPath(@Nullable FileSystem fs,
+                   @Nullable String bucket,
+                   @Nullable String object) {
+        if (bucket == null) {
+            bucket = "";
+        }
+        checkArgument(!bucket.contains("/"),
+                "GCS bucket may not contain a slash");
+        checkArgument(bucket.isEmpty()
+                        || bucket.matches("[a-z0-9][-_a-z0-9.]+[a-z0-9]"),
+                "GCS bucket names must contain only lowercase letters, numbers, "
+                        + "dashes (-), underscores (_), and dots (.). Bucket names "
+                        + "must start and end with a number or letter. "
+                        + "See https://developers.google.com/storage/docs/bucketnaming "
```

Comment:
it seems regex may more rules as per developer-doc
> Bucket names must contain 3 to 63 characters. Names containing dots can contain up to 222 characters, but each dot-separated component can be no longer than 63 characters.
Bucket names cannot be represented as an IP address in dotted-decimal notation (for example, 192.168.5.4).
Bucket names cannot begin with the "goog" prefix.
Bucket names cannot contain "google" or close misspellings of "google".